src/mlpack/tests/kde_test.cpp

```diff
@@ -818,4 +857,378 @@ BOOST_AUTO_TEST_CASE(NotTrained)
   BOOST_REQUIRE_THROW(kde.Evaluate(estimations), std::runtime_error);
 }
 
+/**
+ * Test single KD-tree implementation results against brute force results using
+ * Monte Carlo estimations when possible.
+ */
+BOOST_AUTO_TEST_CASE(GaussianSingleKDTreeMonteCarloKDE)
+{
+  arma::mat reference = arma::randu(2, 3000);
+  arma::mat query = arma::randu(2, 100);
+  arma::vec bfEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+  arma::vec treeEstimations = arma::vec(query.n_cols, arma::fill::zeros);
+  const double kernelBandwidth = 0.5;
+  const double relError = 0.05;
+
+  // Brute force KDE.
+  GaussianKernel kernel(kernelBandwidth);
+  BruteForceKDE<GaussianKernel>(reference,
+                                query,
+                                bfEstimations,
+                                kernel);
+
+  // Optimized KDE.
+  metric::EuclideanDistance metric;
+  KDE<GaussianKernel,
+      metric::EuclideanDistance,
+      arma::mat,
+      tree::KDTree>
+    kde(relError,
+        0.0,
+        kernel,
+        KDEMode::SINGLE_TREE_MODE,
+        metric,
+        true,
+        0.95,
+        100,
+        2,
+        0.7);
+  kde.Train(reference);
+  kde.Evaluate(query, treeEstimations);
+
+  // The Monte Carlo estimation has a random component so it can fail. Therefore
+  // we require a reasonable amount of results to be right.
+  size_t correctResults = 0;
+  for (size_t i = 0; i < query.n_cols; ++i)
+  {
+    const double resultRelativeError =
+      std::abs((bfEstimations[i] - treeEstimations[i]) / bfEstimations[i]);
+    if (resultRelativeError < relError)
+      ++correctResults;
+  }
+
+  BOOST_REQUIRE_GT(correctResults, 70);
```

Comment:
Have you tried running this test many times with different random seeds to check if it is robust?  I don't know how likely it is for the Monte Carlo sampling to fail, although based on what I saw when I ran some simulations earlier, getting 70 out of 100 within the bound is probably a good bound that won't fail.

Reply:
Yes, it seems to always have a reasonable amount of fails. Actually, none of the tests I did ever got less than 70 within the bound :)